### PR TITLE
follow Lmod rules for path updates

### PR DESCRIPTION
**Pull request by gsell**
Changes implemented in this PR:
* add config option `lmod_path_rules`
* if `lmod_path_rules` is set: append/prepend dir to path and remove all other occurrences of dir from path

### CODE REVIEW

File: tcl/envmngt.tcl.in [hunk 1]
```diff
@@ -1815,17 +1815,32 @@ proc add-path {cmd mode dflbhv args} {
...
-         # unsetting it)
-         if {$val ne {} || [info exists countarr()]} {
-            set sep [expr {$val eq $separator ? {} : $separator}]
-            set val [expr {$bhv eq {prepend} ? "$dir$sep$val" :\
-               "$val$sep$dir"}]
+   if {[getConf lmod_path_rule]} {
+      set mpath_list [split $val $separator]
+      foreach dir $path_list {
+         if {$bhv eq {prepend}} {
+            set mpath_list "$dir [lsearch -inline -all -not -exact\
```
Comment (xdelaruelle) [inline tcl/envmngt.tcl.in]: I recommend having the path removal command separated, as it is the same whether we append or we prepend.

File: tcl/envmngt.tcl.in [hunk 2]
```diff
@@ -1815,17 +1815,32 @@ proc add-path {cmd mode dflbhv args} {
...
+                $mpath_list $dir]"
          } else {
-            set val $dir
+            set mpath_list "[lsearch -inline -all -not -exact\
+                $mpath_list $dir] $dir"
```
[the remark this answers is not in the document]
Comment (xdelaruelle) [reply tcl/envmngt.tcl.in]: I am not an Emacs user so I cannot help here unfortunately.

File: tcl/envmngt.tcl.in [hunk 3]
```diff
@@ -1815,17 +1815,32 @@ proc add-path {cmd mode dflbhv args} {
...
+            if {$val ne {} || [info exists countarr()]} {
+               set sep [expr {$val eq $separator ? {} : $separator}]
+               set val [expr {$bhv eq {prepend} ? "$dir$sep$val" :\
+                                  "$val$sep$dir"}]
+            } else {
+               set val $dir
+            }
          }
       }
       if {[info exists countarr($dir)]} {
```
[the remark this answers is not in the document]
Comment (gsell) [reply tcl/envmngt.tcl.in]: Done in the revised code.

File: configure [hunk 4]
```diff
@@ -34,7 +34,7 @@ multilibsupport libdir64 libdir32 versioning silentshdbgsupport \
 setshellstartup quarantinesupport autohandling availindepth implicitdefault \
 extendeddefault moduleshome initconfin pager pageropts verbosity color \
 darkbgcolors lightbgcolors termbg lockedconfigs icase unloadmatchorder \
-searchmatch modulepath loadedmodules quarantinevars wa277 advversspec ml \
+searchmatch modulepath loadedmodules quarantinevars wa277 lmodpath advversspec ml \
```
Comment (xdelaruelle) [inline configure]: need to adapt the option name here

File: tcl/init.tcl.in [hunk 5]
```diff
@@ -111,6 +111,7 @@ array set g_config_defs [list\
       alias indesym sym tag hidden key} {} {} eltlist}\
    list_terse_output {MODULES_LIST_TERSE_OUTPUT {@listterseoutput@} 0 l\
       {header idx variant alias indesym sym tag hidden key} {} {} eltlist}\
+   path_entry_reorder {MODULES_PATH_ENTRY_REORDER 1 0 b {0 1}}\
```
Comment (xdelaruelle) [inline tcl/init.tcl.in]: use `@pathentryreorder@` as default value for option

File: tcl/envmngt.tcl.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (xdelaruelle) [reply tcl/envmngt.tcl.in]: I think code addition would be better located within the `if {[info exists countarr($dir)]} {` code block as there would be no need to handle duplicates, ignore-refcount mode and also no need to have specific code to increase or not the reference counter.

There would be a small duplication of code to add the path entry into the variable value, but everything else will be already managed.

File: tcl/envmngt.tcl.in [hunk 6]
```diff
@@ -1816,27 +1816,46 @@ proc add-path {cmd mode dflbhv args} {
...
+         if {[info exists countarr($dir)]} {
+            # do not increase counter if bare separator string is added or if
+            # reference count is ignored (--ignore-refcount set) unless if
+            # duplicate mode is enabled (--duplicates set)
+            if {!$val_set_is_delim && (!$ign_refcount || $allow_dup)} {
+               incr countarr($dir)
+            }
+         } else {
+            set countarr($dir) 1
+         }
```
[the remark this answers is not in the document]
Comment (xdelaruelle) [reply tcl/envmngt.tcl.in]: Do not hesitate to tell me if it is still not clear.

**Comment by xdelaruelle** [comment]
@gsell Many thanks for the new commits.

* To go further, please follow the [Add new configuration option](https://modules.readthedocs.io/en/latest/devel/add-new-config-option.html) guide to add documentation, completion and tests.

> It will especially help to fix the existing "config" tests to see if the change does not introduce regression.

* Specific test to demonstrate the new feature should also be added. Especially to highlight corner cases.

* All your commits should also be adapted to include a [developer certificate of origin (DCO)](https://modules.readthedocs.io/en/latest/CONTRIBUTING.html#developer-certificate-of-origin-dco
)

**Comment by gsell** [comment]
@xdelaruelle the commits are signed now. But I still have a merge commit which is not signed:
```
commit 03b655a233da6a39bd3ff966ef2caa2de7085168 (origin/main, origin/HEAD, main)
Merge: 6860bdea 7c91bea8
Author: Achim Gsell <achim.gsell@psi.ch>
Date:   Fri Dec 19 09:55:08 2025 +0100

    Merge pull request #1 from envmodules/main
    
    pull changes from upstream
```
and I don't know how to sign it.

Another issue might be:
```
commit fb65540800b0cd46a2fcb267502e2cbf7372d758 (HEAD -> enhancement/Update_order_of_entries_in_path_environment_variable, origin/enhancement/Update_order_of_entries_in_path_environment_variable)
gpg: Signature made Mon Jan  5 17:45:46 2026 CET
gpg:                using EDDSA key 37BB6807BB9F78F76CB16F1F0B5913AD51BFF7DA
gpg: Good signature from "Achim Gsell <achim.gsell@psi.ch>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 37BB 6807 BB9F 78F7 6CB1  6F1F 0B59 13AD 51BF F7DA
Author: Achim Gsell <achim.gsell@psi.ch>
Date:   Thu Dec 18 14:44:27 2025 +0100

    setting default value for path_entry_reorder fixed
    
    Signed-off-by: Achim Gsell <achim.gsell@psi.ch>
```
We are using GPG mainly internally and I (we) signed our commits. Would it be possible that you sign my key?

**Comment by gsell** [comment]
Help with the tests would be very appreciated! It took me a while to understand the design/concept behind them. For me it is still not clear where to add the new tests. I didn't found an example where similar tests are performed after changing something with `module config`.

**Comment by gsell** [comment]
While trying to understand how to run the tests, I run into an issue with resolving group id's: In some cases group id's cannot be resolved to a name. In particular this is the case if you are in an AFS process authentication group (PAG).

I can open a new issue for this problem.

**Comment by xdelaruelle** [comment]
> @xdelaruelle the commits are signed now. But I still have a merge commit which is not signed:
> ...
> and I don't know how to sign it.

No need to sign this one, it will disappear when I will rebase everything. The important is to sign the commit with your content, which is done now (the DCO CI test is green).


> Another issue might be:
> 
> ```
> commit fb65540800b0cd46a2fcb267502e2cbf7372d758 (HEAD -> enhancement/Update_order_of_entries_in_path_environment_variable, origin/enhancement/Update_order_of_entries_in_path_environment_variable)
> gpg: Signature made Mon Jan  5 17:45:46 2026 CET
> gpg:                using EDDSA key 37BB6807BB9F78F76CB16F1F0B5913AD51BFF7DA
> gpg: Good signature from "Achim Gsell <achim.gsell@psi.ch>" [unknown]
> gpg: WARNING: This key is not certified with a trusted signature!
> gpg:          There is no indication that the signature belongs to the owner.
> Primary key fingerprint: 37BB 6807 BB9F 78F7 6CB1  6F1F 0B59 13AD 51BF F7DA
> Author: Achim Gsell <achim.gsell@psi.ch>
> Date:   Thu Dec 18 14:44:27 2025 +0100
> 
>     setting default value for path_entry_reorder fixed
>     
>     Signed-off-by: Achim Gsell <achim.gsell@psi.ch>
> ```
> 
> We are using GPG mainly internally and I (we) signed our commits. Would it be possible that you sign my key?

I have no GPG key. Crypto signature is not required here. Just the "Signed-off-by: " line is needed.

**Comment by xdelaruelle** [comment]
> While trying to understand how to run the tests, I run into an issue with resolving group id's: In some cases group id's cannot be resolved to a name. In particular this is the case if you are in an AFS process authentication group (PAG).
> 
> I can open a new issue for this problem.

Yes feel free to open an issue for that. What will be important is to describe it for me to be able to reproduce the issue on my side.

**Comment by xdelaruelle** [comment]
Many thanks @gsell for your contribution. I will handle the rest (rebase, commit squash and fixing external tests). Everything should be merged in the coming days.

**Comment by xdelaruelle** [comment]
I have rebased branch, reorganized commits and extend a bit the documentation with an additional commit.

The CI test case where `path_entry_reorder` is enabled by default raises errors. I will investiguate that (tests should be adapted or some corner cases are unveiled by these tests).
